### ISSUE TITLE
High memory usage 0.12.2

### ISSUE THREAD

**Issue by freeseacher** [influxdata/influxdb#6513]
### Bug report

__System info:__ [ version 0.12.2, rhel 7.2]

__Steps to reproduce:__

1. start system
2. wait for system started that's is about 4 minutes. 
3. start pooler. 
4. got 

`ps wavx | grep influx`

` 6940 ?        Sl    25:19 109030  5031 120035932 *74729508* 75.6 /usr/bin/influxd -config /etc/influxdb/influxdb.conf`

__Expected behavior:__  memory usage under 20G
__Actual behavior:__ memory usage above 75G
with 
`go tool pprof /bin/influxd http://localhost:8086/debug/pprof/heap`
got result in file
[usage.zip](https://github.com/influxdata/influxdb/files/242936/usage.zip)
my config is just like default one
[config.zip](https://github.com/influxdata/influxdb/files/242949/config.zip)

stats and others
[stats.zip](https://github.com/influxdata/influxdb/files/242953/stats.zip)
[logs.zip](https://github.com/influxdata/influxdb/files/242958/logs.zip)

```
free -m
              total        used        free      shared  buff/cache   available
Mem:          96507       82178       13470          96         858       13564
Swap:         54271       16075       38196
```

Please help to understand why memory usage so extremely high.

**Comment by jwilder** [comment]
What does `free -m` show?

**Comment by jwilder** [comment]
It looks like you have about 5M series.  Are you running any queries?  If so, what are they?

**Comment by freeseacher** [comment]
no. i just adding new metrics. 
after several minutes of running pooler i start got such lines in logs
```
2016-04-29 21:14:14,939 [discovery] Failed to spool collected metrics to 10.36.129.4:8086: HTTP 599: Failed connect to 10.36.129.4:8086; Connection refused
```

**Comment by jwilder** [comment]
Leaking connections possibly?  Are the your writes batched?  How big are the batches?  How many writers?

**Comment by freeseacher** [comment]
oh i see. in that moment i have a stacktrace in my influx logs 
[logs.txt](https://github.com/influxdata/influxdb/files/242988/logs.txt)

**Comment by jwilder** [comment]
What is your writer process `[discovery]`?  If it's written in go and use the v2 client, there was a bug about leaked connections that was recently fixed.

**Comment by freeseacher** [comment]
no. Its written in python and i have lots of them. 
4 servers with each with 18 process with 187 threads. not every thread works with metrics of cause. 
i am unsure about batch size. but of cause it is batched.

**Comment by dvolodin7** [comment]
discovery process uses libcurl to spool the metrics. Each discovery process (72 at all) spools all collected metrics every 250ms in a single batch.

At this time we collect ~100 metrics from 50k objects every 300 seconds and have the plans to increase amount of objects up to 250k

**Comment by freeseacher** [comment]
ok. looks like we found and fixed problem with our discovery process.

i am unsure weather i have to open another issue or not. 
but problem looks the same. high memory usage. 
![pic](https://cloud.githubusercontent.com/assets/289277/14943648/01df4190-0fe7-11e6-95a5-f44f3b50232f.PNG)

but probably the reason is different. 
[the_log.txt](https://github.com/influxdata/influxdb/files/244273/the_log.txt)
it looks like while compacting the database influx eats all the memory 

then after restart. it begun to compact db again, at the moment influx is still not able to accept metrics. So 
1. restart time is awful 
2. compact db east all the memory. 

i think the first problem is in architecture and can't be fixed in soon, but can i handle the second one somehow ?

**Comment by jwilder** [comment]
The restart time is a known issue with some datasets.  Issue to follow is: #6250.  There is lock contention when reloading the in-memory index that slows some datasets down a lot.

For the compactions, are you overwriting points or writing to series in the past?   Can you provide some sample data that you're writing?

**Comment by whitelynx** [comment]
Possible duplicate of #6243?

**Comment by jwilder** [comment]
@freeseacher When your heap starts to spike, would you be able to grab a snapshot of the heap and goroutines and attach the output here?

```
curl -o heap.txt http://localhost:8086/debug/pprof/heap?debug=1
curl -o goroutine.txt http://localhost:8086/debug/pprof/goroutine?debug=1
```

**Comment by jbye** [comment]
Not to hijack the issue, but I am also experiencing very high memory consumption. Influx will essentially eat all the RAM on the machine (CentOS 7.1) within ~ 1 minute after start. It also takes a long while to start -- it freezes up completely while loading the database index (some which take almost 3 minutes).

I have attached the heap and goroutine from my system.

[goroutine.txt](https://github.com/influxdata/influxdb/files/248178/goroutine.txt)
[heap.txt](https://github.com/influxdata/influxdb/files/248179/heap.txt)

**Comment by jwilder** [comment]
#6618 should help start up time some.

@freeseacher Are you able to grab a heap and profile using the commands above when you heap starts to spike?

**Comment by jwilder** [comment]
Should be fixed via #6653 and #6743.  If there are still issues, please let us know.
